File: docs/kb/tools/topic_index.js
#!/usr/bin/env node
// Generate docs/kb/topic-index.md. Node port of topic_index.py — same algorithm,
// byte-identical output. See topic_index.py for why the index exists and why there
// is more than one implementation.
//
//   node docs/kb/tools/topic_index.js [--check]
//
// Prefer the dispatcher, which picks whatever runtime is present:
//   docs/kb/tools/topic-index.sh [--check]

'use strict';
const fs = require('fs');
const path = require('path');

const KB_DIR = 'docs/kb';
const OUT_FILE = path.join(KB_DIR, 'topic-index.md');
const TOP_TERMS = 18;

const STOP = new Set(`the and that this with for not but are was were has have had which from into
when what where how why it its a an of to in on is as at by or be so if than then there here one
two can cannot does doing done would should could must will now still only also any all every each
same other more most less just like who whom they them their you your we our us i me my he she his
her him no yes doc docs md see note notes used using use uses new old first second test tests code
line lines file files case cases thing things way ways well much many some such over under after
before because since while`.split(/\s+/));

// ASCII-only lowercasing on purpose — see the note in topic_index.py.
const asciiLower = (s) => s.replace(/[A-Z]/g, (c) => String.fromCharCode(c.charCodeAt(0) + 32));

function walk(dir) {
  const out = [];
  for (const e of fs.readdirSync(dir, { withFileTypes: true })) {
    const p = path.join(dir, e.name);
    if (e.isDirectory()) out.push(...walk(p));
    else if (e.isFile() && e.name.endsWith('.md') &&
             e.name !== 'README.md' && e.name !== 'topic-index.md') out.push(p);
  }
  return out;
}

const docs = () =>
  walk(KB_DIR).map((p) => path.relative(KB_DIR, p).split(path.sep).join('/')).sort();

const read = (rel) => fs.readFileSync(path.join(KB_DIR, rel), 'utf8');

function headings(text) {
  const out = [];
  for (const line of text.split('\n')) {
    const m = /^#{2,3}\s+(.+?)\s*$/.exec(line);
    if (!m) continue;
    const h = m[1]
      .replace(/`([^`]*)`/g, '$1')
      .replace(/\[([^\]]*)\]\([^)]*\)/g, '$1')
      .replace(/[*_]{1,2}/g, '')
      .trim();
    if (h && !out.includes(h)) out.push(h);
  }
  return out;
}

function words(text) {
  return asciiLower(text.replace(/```[\s\S]*?```/g, ' ').replace(/`[^`]*`/g, ' '))
    .split(/[^a-z0-9-]+/)
    .filter((w) => w.length >= 3 && !STOP.has(w));
}

function frequencies(items) {
  const f = new Map();
  for (const i of items) f.set(i, (f.get(i) || 0) + 1);
  return f;
}

function distinctiveTerms(freqs, nDocs, rel, k = TOP_TERMS) {
  const tf = freqs.get(rel);
  const scored = [];
  for (const [w, c] of tf) {
    let df = 0;
    for (const other of freqs.values()) if (other.has(w)) df++;
    scored.push([w, c * Math.log(nDocs / Math.max(1, df))]);
  }
  scored.sort((a, b) => (b[1] - a[1]) || (a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0));
  return scored.slice(0, k).map((p) => p[0]).sort();
}

const termsFromName = (rel) => {
  const stem = rel.endsWith('.md') ? rel.slice(0, -3) : rel;
  return [stem, stem.replace(/-/g, ' ')];
};

function render() {
  const ds = docs();
  const texts = new Map(ds.map((rel) => [rel, read(rel)]));
  const freqs = new Map(ds.map((rel) => [rel, frequencies(words(texts.get(rel)))]));
  const n = ds.length;

  const head =
    '# Topic index — what has already been looked at\n\n' +
    '**GENERATED — do not edit.** `docs/kb/tools/topic-index.sh`\n\n' +
    '## Grep the corpus first\n\n' +
    '```\ngrep -ril "<term>" docs/kb/\n```\n\n' +
    '**That is the search.** This file is for orientation — what each document is about, and\n' +
    'which one owns a topic — not for recall. Measured against fourteen realistic queries the\n' +
    'corpus answered **all fourteen**; this index answered **nine**. It cannot match multi-word\n' +
    'phrases (`import conflict`, `spell list`) because it is built from single words, and a\n' +
    'topic mentioned once loses its place to one discussed throughout. Use it to find the right\n' +
    'document, then read that document; use grep to find out whether anyone has been there.\n\n' +
    'Each document is listed with its filename (hyphenated **and** spaced, because queries are\n' +
    'typed with spaces), the words that most distinguish it from the rest of the corpus, and\n' +
    'every section heading it contains.\n\n' +
    '---\n\n';

  const blocks = ds.map((rel) =>
    `## ${rel}\n\n` +
    `_${termsFromName(rel).join(' · ')}_\n\n` +
    `**topics:** ${distinctiveTerms(freqs, n, rel).join(', ')}\n\n` +
    headings(texts.get(rel)).map((h) => `- ${h}`).join('\n') + '\n');

  return head + blocks.join('\n') + '\n';
}

function main(argv) {
  if (!fs.existsSync(KB_DIR)) {
    process.stderr.write(`error: ${KB_DIR} not found — run from the repository root\n`);
    return 2;
  }
  const content = render();
  const ds = docs();
  const nHeadings = ds.reduce((a, rel) => a + headings(read(rel)).length, 0);

  if (argv.includes('--check')) {
    const current = fs.existsSync(OUT_FILE) ? fs.readFileSync(OUT_FILE, 'utf8') : null;
    if (current === content) {
      console.log(`topic-index.md is up to date (${ds.length} documents, ${nHeadings} headings)`);
      return 0;
    }
    process.stderr.write(
      'topic-index.md is OUT OF DATE — regenerate with docs/kb/tools/topic-index.sh\n');
    return 1;
  }

  fs.writeFileSync(OUT_FILE, content);
  console.log(`wrote ${OUT_FILE} (${ds.length} documents, ${nHeadings} headings)`);
  return 0;
}

process.exit(main(process.argv.slice(2)));

File: docs/kb/tools/topic_index.py
#!/usr/bin/env python3
"""Generate docs/kb/topic-index.md — a flat, greppable map from TOPIC to document.

Port of dev/orcpub/topic_index.clj. Same algorithm, same output; see that file for
why the index exists at all.

Why a port. The original is invoked as `lein with-profile +tools run -m
orcpub.topic-index`, and Leiningen is not installed in the containers agents actually
run in — so the one tool meant to keep the KB navigable was the one nobody could run.
The index went stale for that reason. The script needs nothing but the standard
library, in any language: it reads markdown and counts words.

Run it with the dispatcher, which picks whatever runtime is present:

    docs/kb/tools/topic-index.sh            # regenerate
    docs/kb/tools/topic-index.sh --check    # exit 1 if the index is out of date

Or directly:  python3 docs/kb/tools/topic_index.py

Two deliberate differences from the original Clojure:

  * Nested documents work. The original called .getName on a recursive file-seq,
    discarding the directory, then read docs/kb/<basename> — so it threw
    FileNotFoundException on anything under docs/kb/rescued/. Documents are keyed by
    their path relative to docs/kb.

  * Ties break deterministically. The original sorted by score alone, leaving equal
    scores in hash-map order, so two runs could disagree. Ties break on the word.
"""

import math
import re
import sys
from pathlib import Path

KB_DIR = Path("docs/kb")
OUT_FILE = KB_DIR / "topic-index.md"
TOP_TERMS = 18

STOP = {
    "the", "and", "that", "this", "with", "for", "not", "but", "are", "was", "were", "has", "have",
    "had", "which", "from", "into", "when", "what", "where", "how", "why", "it", "its", "a", "an",
    "of", "to", "in", "on", "is", "as", "at", "by", "or", "be", "so", "if", "than", "then", "there",
    "here", "one", "two", "can", "cannot", "does", "doing", "done", "would", "should", "could",
    "must", "will", "now", "still", "only", "also", "any", "all", "every", "each", "same", "other",
    "more", "most", "less", "just", "like", "who", "whom", "they", "them", "their", "you", "your",
    "we", "our", "us", "i", "me", "my", "he", "she", "his", "her", "him", "no", "yes", "doc", "docs",
    "md", "see", "note", "notes", "used", "using", "use", "uses", "new", "old", "first", "second",
    "test", "tests", "code", "line", "lines", "file", "files", "case", "cases", "thing", "things",
    "way", "ways", "well", "much", "many", "some", "such", "over", "under", "after", "before",
    "because", "since", "while",
}

HEADING_RE = re.compile(r"^#{2,3}\s+(.+?)\s*$")
FENCED_RE = re.compile(r"```.*?```", re.DOTALL)
INLINE_CODE_RE = re.compile(r"`[^`]*`")
WORD_SPLIT_RE = re.compile(r"[^a-z0-9-]+")

# ASCII-only lowercasing on purpose. str.lower() is Unicode-aware and the Clojure
# original used str/lower-case, which is LOCALE-aware -- on a Turkish JVM it folds
# "I" to dotless "i" and silently produces a different index. Document text is being
# tokenised, not read aloud. See docs/kb/locale-safety.md.
ASCII_LOWER = str.maketrans("ABCDEFGHIJKLMNOPQRSTUVWXYZ", "abcdefghijklmnopqrstuvwxyz")


def docs():
    """Every KB document, keyed by path relative to docs/kb, sorted."""
    return sorted(
        str(p.relative_to(KB_DIR))
        for p in KB_DIR.rglob("*.md")
        if p.is_file() and p.name != "README.md" and p.name != "topic-index.md"
    )


def read(rel):
    return (KB_DIR / rel).read_text(encoding="utf-8")


def headings(text):
    """Section headings (## and ###), stripped of inline code, links and emphasis."""
    out = []
    for line in text.splitlines():
        m = HEADING_RE.match(line)
        if not m:
            continue
        h = m.group(1)
        h = re.sub(r"`([^`]*)`", r"\1", h)
        h = re.sub(r"\[([^\]]*)\]\([^)]*\)", r"\1", h)
        h = re.sub(r"[*_]{1,2}", "", h)
        h = h.strip()
        if h and h not in out:
            out.append(h)
    return out


def words(text):
    """Topic words: code stripped, lowercased, >=3 chars, stopwords removed."""
    text = FENCED_RE.sub(" ", text)
    text = INLINE_CODE_RE.sub(" ", text)
    text = text.translate(ASCII_LOWER)
    return [w for w in WORD_SPLIT_RE.split(text) if len(w) >= 3 and w not in STOP]


def frequencies(items):
    f = {}
    for i in items:
        f[i] = f.get(i, 0) + 1
    return f


def distinctive_terms(freqs, n_docs, rel, k=TOP_TERMS):
    """Crude TF-IDF: the words characterising ONE document against the corpus."""
    tf = freqs[rel]
    df = {}
    for w in tf:
        df[w] = sum(1 for other in freqs.values() if w in other)
    scored = [(w, c * math.log(n_docs / max(1, df[w]))) for w, c in tf.items()]
    scored.sort(key=lambda p: (-p[1], p[0]))
    return sorted(w for w, _ in scored[:k])


def terms_from_name(rel):
    stem = rel[:-3] if rel.endswith(".md") else rel
    return [stem, stem.replace("-", " ")]


def render():
    ds = docs()
    texts = {rel: read(rel) for rel in ds}
    freqs = {rel: frequencies(words(t)) for rel, t in texts.items()}
    n = len(ds)

    parts = [
        "# Topic index — what has already been looked at\n\n",
        "**GENERATED — do not edit.** `docs/kb/tools/topic-index.sh`\n\n",
        "## Grep the corpus first\n\n",
        '```\ngrep -ril "<term>" docs/kb/\n```\n\n',
        "**That is the search.** This file is for orientation — what each document is about, and\n"
        "which one owns a topic — not for recall. Measured against fourteen realistic queries the\n"
        "corpus answered **all fourteen**; this index answered **nine**. It cannot match multi-word\n"
        "phrases (`import conflict`, `spell list`) because it is built from single words, and a\n"
        "topic mentioned once loses its place to one discussed throughout. Use it to find the right\n"
        "document, then read that document; use grep to find out whether anyone has been there.\n\n",
        "Each document is listed with its filename (hyphenated **and** spaced, because queries are\n"
        "typed with spaces), the words that most distinguish it from the rest of the corpus, and\n"
        "every section heading it contains.\n\n",
        "---\n\n",
    ]

    blocks = []
    for rel in ds:
        block = (
            f"## {rel}\n\n"
            f"_{' · '.join(terms_from_name(rel))}_\n\n"
            f"**topics:** {', '.join(distinctive_terms(freqs, n, rel))}\n\n"
            + "\n".join(f"- {h}" for h in headings(texts[rel]))
            + "\n"
        )
        blocks.append(block)

    return "".join(parts) + "\n".join(blocks) + "\n"


def main(argv):
    check = "--check" in argv
    if not KB_DIR.is_dir():
        sys.stderr.write(f"error: {KB_DIR} not found — run from the repository root\n")
        return 2

    content = render()
    ds = docs()
    n_headings = sum(len(headings(read(rel))) for rel in ds)

    if check:
        current = OUT_FILE.read_text(encoding="utf-8") if OUT_FILE.exists() else None
        if current == content:
            print(f"topic-index.md is up to date ({len(ds)} documents, {n_headings} headings)")
            return 0
        sys.stderr.write(
            "topic-index.md is OUT OF DATE — regenerate with docs/kb/tools/topic-index.sh\n"
        )
        return 1

    OUT_FILE.write_text(content, encoding="utf-8")
    print(f"wrote {OUT_FILE} ({len(ds)} documents, {n_headings} headings)")
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))
